// console-table-printer/dist/src/utils/table-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultHeaderAlignment = exports.defaultRowAlignment = exports.defaultHeaderFontColor = exports.defaultRowFontColor = exports.colors = exports.alignments = exports.DEFAULT_TABLE_STYLE = void 0;
exports.DEFAULT_TABLE_STYLE = {
    /*
        Default Style
        ┌────────────┬─────┬──────┐
        │ foo        │ bar │ baz  │
        │ frobnicate │ bar │ quuz │
        └────────────┴─────┴──────┘
        */
    headerTop: {
        left: '┌',
        mid: '┬',
        right: '┐',
        other: '─',
    },
    headerBottom: {
        left: '├',
        mid: '┼',
        right: '┤',
        other: '─',
    },
    tableBottom: {
        left: '└',
        mid: '┴',
        right: '┘',
        other: '─',
    },
    vertical: '│',
};
exports.alignments = ['right', 'left', 'center'];
exports.colors = [
    'red',
    'green',
    'yellow',
    'white',
    'blue',
    'magenta',
    'cyan',
    'crimson',
    'white_bold',
    'reset',
];
exports.defaultRowFontColor = 'white';
exports.defaultHeaderFontColor = 'white_bold';
exports.defaultRowAlignment = 'right';
exports.defaultHeaderAlignment = 'center';
